python/ray/util/worker_group.py

```diff
@@ -0,0 +1,308 @@
+            ]
+            # Wait for actors to die gracefully.
+            done, not_done = ray.wait(done_refs, timeout=patience_s)
+            if not_done:
+                logger.debug("Graceful termination failed. Falling back to "
+                             "force kill.")
+                # If all actors are not able to die gracefully, then kill them.
+                for worker in self.workers:
+                    ray.kill(worker.actor)
+
+        logger.debug("Shutdown successful.")
+        self.workers = []
+
+    def execute_async(self, func: Callable[..., T], *args,
+                      **kwargs) -> List[ObjectRef]:
+        """Execute ``func`` on each worker and return the futures.
+
+        Args:
+            func (Callable): A function to call on each worker.
+            args, kwargs: Passed directly into func.
+
+        Returns:
+            (List[ObjectRef]) A list of ``ObjectRef`` representing the
+                output of ``func`` from each worker. The order is the same
+                as ``self.workers``.
+
+        """
+        if len(self.workers) <= 0:
+            raise RuntimeError("There are no active workers. This worker "
+                               "group has most likely been shut down. Please"
+                               "create a new WorkerGroup or restart this one.")
+
+        return [
+            w.actor._BaseWorkerMixin__execute.remote(func, *args, **kwargs)
+            for w in self.workers
+        ]
+
+    def execute(self, func: Callable[..., T], *args, **kwargs) -> List[T]:
+        """Execute ``func`` on each worker and return the outputs of ``func``.
+
+        Args:
+            func (Callable): A function to call on each worker.
+            args, kwargs: Passed directly into func.
+
+        Returns:
+            (List[T]) A list containing the output of ``func`` from each
+                worker. The order is the same as ``self.workers``.
+
+        """
+        return ray.get(self.execute_async(func, *args, **kwargs))
+
+    def execute_single_async(self, worker_index: int, func: Callable[..., T],
+                             *args, **kwargs) -> ObjectRef:
+        """Execute ``func`` on worker ``worker_index`` and return futures.
+
+        Args:
+            worker_index (int): The index to execute func on.
+            func (Callable): A function to call on the first worker.
+            args, kwargs: Passed directly into func.
+
+        Returns:
+            (ObjectRef) An ObjectRef representing the output of func.
+
+        """
+        if worker_index >= len(self.workers):
+            raise ValueError(f"The provided worker_index {worker_index} is "
+                             f"not valid for {self.num_workers} workers.")
+        return self.workers[worker_index].actor._BaseWorkerMixin__execute\
+            .remote(
+            func, *args, **kwargs)
+
+    def execute_single(self, worker_index: int, func: Callable[..., T], *args,
```

Comment:
Oh, I didn't realize that we were dynamically shipping the `func` here. That doesn't seem like what we want to do in the common case -- can we make the interface to execute a specific method on the actor and then SGD happens to do dynamic function shipping?

Reply:
Yep, will change this to require an `actor_cls` in `__init__`, and move this logic to be SGD-specific.